### ISSUE TITLE
Pagination broke after e6357bc

### ISSUE THREAD

**Issue by RobertLowe** [marmelab/react-admin#5168]
After: https://github.com/marmelab/react-admin/commit/e6357bcc95f53d588a8b44073c1dc22fcbcc9b85
Specifically: https://github.com/marmelab/react-admin/blob/master/packages/ra-core/src/controller/useListController.ts#L187-L189

Broke my apps pagination. I wasn't able to investigate too deeply but I think the root cause may have been that I mutate params inside it's `buildQuery`, so I think there may be a cache mismatch or something. `defaultTotal` was correct, `total` was undefined in the new change, forcing the page to continually `setPage(1)` below.

**What you were expecting:**
No breaking change from 3.7.1 to 3.8.1

**What happened instead:**
Pagination broke

**Steps to reproduce:**
I can't give a full reproduction presently, but a `dataProvider`'s `buildQuery` which mutates `params` of a `GET_LIST`

**Related code:**
Specifically: https://github.com/marmelab/react-admin/blob/master/packages/ra-core/src/controller/useListController.ts#L187-L189

**Environment**

* React-admin version: 3.8.1
* Last version that did not exhibit the issue (if applicable): 3.7.1
* React version: ^16.9.0
* Browser: Chrome

**Comment by fzaninotto** [comment]
>  the root cause may have been that I mutate params inside its buildQuery

Shall I deduce that you are using GraphQL? Can you give more details about your dataProvider?

also, mutating params is never a good idea. Are you forced to do that?

Finally, without a way to reproduce the issue, we can't investigate it further. I don't see what's wrong with the code snippet you've mentioned.

**Comment by RobertLowe** [comment]
Yes, sorry I didn't mention that. I'm using graphql.

Can you elaborate on why mutating params isn't?

Here's a snippet:

The buildQuery looks like this:

```javascript
export const queryBuilder = introspectionResults => (aorFetchType, resourceName, params) => {
  const resource = introspectionResults.resources.find(r => r.type.name === resourceName);
  let result = {};
  switch (aorFetchType) {
   case 'GET_LIST':
   case 'GET_MANY':
   case 'GET_MANY_REFERENCE':
    params.page = params.pagination.page;
    params.perPage = params.pagination.perPage;

    if(params.sort && params.sort.field){
      if(params.sort.field === "id")
        params.sort.field = "_id";
      params.sort = `${(params.sort.field).toUpperCase()}_${params.sort.order.toUpperCase()}`
    }

    result =  {
      query: gql`query ${inflection.camelize(resource[aorFetchType].name, true)}($filter: FilterFindMany${resourceName}Input, $sort: SortFindMany${resourceName}Input, $page: Int, $perPage: Int) {
        data: ${inflection.camelize(resource[aorFetchType].name, true)}(filter: $filter, sort: $sort, page: $page, perPage: $perPage) {
          items {
            id: _id
            ${buildFieldList(introspectionResults, resource, aorFetchType, resource)}
          }
          total: count
        }
      }`,
      variables: params,
      parseResponse: response => ({
        data: response.data.data.items,
        total: response.data.data.total
      })
    };
    break;
  default:
    return undefined;
  }
  return result;
}
```

You can see here, I'm mapping the the param variables differently than passed it. This code works fine in 3.7.1, but in 3.8.1 it no longer works as `total` become undefined briefly

I'll try to make some time to produce a reproduction but it may be a while before I can 2-3 weeks.

For more context, I've written a dataProvider for [graphql-compose-mongoose](https://github.com/graphql-compose/graphql-compose-mongoose), here's an example of what pagination looks like: https://graphql-compose.herokuapp.com/user/?query=%0A%7B%0A%20%20userPagination(filter%3A%20%7B%20gender%3A%20male%20%7D%2C%20perPage%3A%202%2C%20page%3A%202%2C%20sort%3A%20_ID_ASC)%20%7B%0A%20%20%20%20items%20%7B%0A%20%20%20%20%20%20name%0A%20%20%20%20%20%20languages%20%7B%0A%20%20%20%20%20%20%20%20language%0A%20%20%20%20%20%20%20%20skill%0A%20%20%20%20%20%20%7D%0A%20%20%20%20%20%20contacts%20%7B%0A%20%20%20%20%20%20%20%20email%0A%20%20%20%20%20%20%7D%0A%20%20%20%20%20%20gender%0A%20%20%20%20%20%20age%0A%20%20%20%20%20%20salaryDecimal%0A%20%20%20%20%7D%0A%20%20%20%20count%0A%20%20%20%20pageInfo%20%7B%0A%20%20%20%20%20%20currentPage%0A%20%20%20%20%20%20perPage%0A%20%20%20%20%20%20itemCount%0A%20%20%20%20%20%20pageCount%0A%20%20%20%20%20%20hasPreviousPage%0A%20%20%20%20%20%20hasNextPage%0A%20%20%20%20%7D%0A%20%20%7D%0A%7D%0A%20%20%20%20%20%20

And more fun examples if you're interested: https://graphql-compose.herokuapp.com/

Sorry I can't post a full example presently.

**Comment by fzaninotto** [comment]
As in JavaScript, objects are passed by reference, mutating params mutates the original object and may have nasty side effects. 

What if you start by cloning the params before passing them to the query? something like:

```diff
export const queryBuilder = introspectionResults => (aorFetchType, resourceName, params) => {
+ const myParams = { ...params }; // clone params
  const resource = introspectionResults.resources.find(r => r.type.name === resourceName);
  let result = {};
  switch (aorFetchType) {
   case 'GET_LIST':
   case 'GET_MANY':
   case 'GET_MANY_REFERENCE':
-   params.page = params.pagination.page;
-   params.perPage = params.pagination.perPage;
+   myParams .page = params.pagination.page;
+   myParams .perPage = params.pagination.perPage;

-   if(params.sort && params.sort.field){
-     if(params.sort.field === "id")
-       params.sort.field = "_id";
-     params.sort = `${(params.sort.field).toUpperCase()}_${params.sort.order.toUpperCase()}`
-   }
+   if(params.sort && params.sort.field){
+     if(params.sort.field === "id")
+       myParams.sort.field = "_id";
+     myParams.sort = `${(params.sort.field).toUpperCase()}_${params.sort.order.toUpperCase()}`
+   }

    result =  {
      query: gql`query ${inflection.camelize(resource[aorFetchType].name, true)}($filter: FilterFindMany${resourceName}Input, $sort: SortFindMany${resourceName}Input, $page: Int, $perPage: Int) {
        data: ${inflection.camelize(resource[aorFetchType].name, true)}(filter: $filter, sort: $sort, page: $page, perPage: $perPage) {
          items {
            id: _id
            ${buildFieldList(introspectionResults, resource, aorFetchType, resource)}
          }
          total: count
        }
      }`,
-     variables: params,
+     variables: myParams,
      parseResponse: response => ({
        data: response.data.data.items,
        total: response.data.data.total
      })
    };
    break;
  default:
    return undefined;
  }
  return result;
}
```

**Comment by RobertLowe** [comment]
Hi,

I investigated the issue, cloning params does not solve the problem.

**Comment by djhi** [comment]
We need a way to reproduce the issue to investigate. Please provide either a codesandbox (preferably) or a github repository.

**Comment by RobertLowe** [comment]
@djhi okay, reproduction made: https://github.com/RobertLowe/react-admin-5168-reproduction

`npm install`
`npm start`

1. Change page to limit to 5.
2. Try to change page.

**Comment by RobertLowe** [comment]
* example works on codesandbox: http://codesandbox.io/s/github/RobertLowe/react-admin-5168-reproduction/tree/master/?file=/src/index.js

**Comment by fzaninotto** [comment]
Thanks for the codesandbox, I can reproduce the problem: when changing the number of results per page to 5, and asking for page 2, the app properly makes a request for page 2, but immediately after, it makes a request for page 1. I'm not sure this is linked to GraphQL though. Will investigate further.

**Comment by Blagoj5** [comment]
Same thing is happening in my app and im just using normal REST api with express as a backend. What  could be the problem? I'm seeing the network panel in chrome dev tools, first it fetches ?page=2 and immedietly after ?page=1. I tried implementing custom pagination where i use useListContext hook and i even setPage(2) from there but still same issue. I was reading somewhere that you must have CONTENT RANGE in the header is that true?

**Comment by RobertLowe** [comment]
Bump, any ETA on this?

**Comment by toti1212** [comment]
I'm interested in this solution. I had the same issue but without GraphQL. Passing through the pages, sometimes jump into 1 again. I couldn't detect a pattern.

**Comment by djhi** [comment]
It's probably related to #5185

**Comment by fzaninotto** [comment]
Fixed by #5263
